### ISSUE TITLE
Prefer `Delegate` over `Signal`

### ISSUE THREAD

**Issue by DanRStevens** [OutpostUniverse/OPHD#875]
We should come up with a convention for `SignalSource` accessor names.

I feel like `getClickSignalSource()` is a bit too wordy. Using just the signal name such as `click()` seems a bit too brief, and may be mistaken for the signal handler.

Example:
```cpp
NAS2D::SignalSource<>& getClickSignalSource()
{
    return mClickSignal;
}
```

**Comment by belgianguy** [comment]
In my opinion the wordy-ness of functions should be inversely proportional to their ease of use, domain knowledge present in contributors and its available documentation. Once these above concepts are improved upon to a certain level, wordyness can be reduced.

For the Signal and SignalSource principles, all above currently rate "Low" for me personally.

I only remember Qt having some sort of the same functionality, but they too have different conventions that may or may not apply here, so I try to see it as a NAS2D concept currently, nothing else, to avoid inferring wrong ideas and conventions.
As we have seen misuse and/or misunderstanding of the concepts in the current codebase, I do think there is an argument to be made to use the "above average" wordyness for now. Especially to avoid the pitfalls of making them look like event handlers when they are a completely different beast is adding another point for me to keep using verbose function names.

What would also help:
* NAS2D documentation about the Signal and SignalSource principles (http://nas2d.lairworks.com/nas2d_doc/ ?)
* While Doxygen is fine to get an idea of what a specific function does, for event handling it's always nice to see a full fledged example with a clear flow of data from action to reaction. (e.g. Click a button and see a square change color)
* Simple and straightforward Do's and Don't example with a Button, but a "Don't" example is hard to showcase in the OPHD codebase. Perhaps a "Pitfalls" section here on GitHub can be considered with a :heavy_check_mark: and a :heavy_multiplication_x: example?
* A pristine go-to golden standard example in the production code of OPHD that can be used for others to take inspiration from and which can be referred to for refactoring needs if code smells start to appear. This too could be listed in the documentation somewhere to prevent others from taking inspiration from bad/old logic and perpetuating wrong implementations.

Just my 2.00 ç

**Comment by DanRStevens** [comment]
Makes sense.

I believe Leeor mentioned Qt Signals and Slots before, so the current system may have taken inspiration from that. It's been years since I've looked at Qt code though, so I'm no longer familiar with it. That might be a good place to look though for inspiration on naming conventions.

I think it makes sense to design for new developers with no prior familiarity with the code base. Seems like a good idea to try and maximize who will be able to contribute to the project. Plus, I kind of figure easy to read code is generally a sign of good well designed code. I think we still have a ways to go on improvements in that direction. On the plus side, there has been some pretty major refactoring in the last couple of years, so we at least seem to be on the right trajectory.

----

Wow, that documentation hosted on the NAS2D website is ancient. It's probably years out of date. So much has changed since then.

I opened an issue to updated or remove the hosted documentation:
https://github.com/lairworks/nas2d-core/issues/860

I also opened a new issue about improving the documentation. It's been neglected for years, and I've often found much of it to be pretty low value, mostly just restating the class name. We can definitely do better. Actually, that may be an area we could use help with, especially from someone without pre-conceived notions of how the code is supposed to work. A collection of questions or areas of uncertainty may help spur the needed documentation.
https://github.com/lairworks/nas2d-core/issues/861

A bit of good news, I updated the Doxygen config file a few days ago, and made a few updates to the documentation style. You can regenerate the documentation locally by running `doxygen` from the NAS2D folder. It produces a `doc/` folder (ignored by `.gitignore`).

We should probably document how to generate the documentation as part of the ReadMe. Speaking of which, the main ReadMe probably needs a bit of attention, which I've just opened an issue for:
#878

**Comment by DanRStevens** [comment]
I check the Qt docs on Signal and Slots:
https://doc.qt.io/qt-5/signalsandslots.html

They design their system in such a way that there is no named object to handle the set of connections. Rather they have a global function which is called with both source and receiver to form the connections. As such, they have no pattern for names corresponding to `Signal` or `SignalSource`.

**Comment by ldicker83** [comment]
I tend to not use get/set names for things, and opt to have the return types and function parameters make that known.

E.g., instead of `getClickSignal` and `setClickSignal`, it would be `reftype& clickSignal()` and `void clickSignal(reftype& const)`. That's the convention I've used throughout the code.

A lot of the really short names actually come from Win32 GUI programming -- e.g., onMouseDown, onClick, etc. The even shorter names are shorthand for me but we've seen how that can be misunderstood very easily so I'd prefer something like what I suggested above.

**Comment by DanRStevens** [comment]
I came across an interesting design quirk of `MapViewState` that makes it behave unlike other classes that have been converted from using `Signal` to `Delegate`, and so may require a different approach.

Summary:
When using an init list to construct a member field and pass in a `Delegate` which points to a method on the container object, the contained object must never call that `Delegate` from the contained object's constructor (or anything called by it's constructor).

Calling such a container object method through a `Delegate` is an object construction order problem, and will result in undefined behaviour, likely leading to a crash. The problem is contained objects (member fields of the container object) must be fully constructed before the container object is fully constructed. That means during the constructor of a contained object, the container is not yet fully constructed. Hence calling any methods on the container from the contained object constructor would result in undefined behaviour.

Normally there isn't yet a reference to an object that isn't fully constructed, except within the constructor of the object itself, using the `this` pointer. As that is exactly what tends to be captured in a `Delegate`, we've essentially passed a copy of `this` to a still constructing object outside of the object to be used elsewhere. It's fine to hold such a copy of `this`, so long as we don't use the copy of `this` from outside the object before the object finishes construction. Before construction completes, values may still be uninitialized, and class invariants have not yet been established.

It should be noted that a common use for `Delegate` is to send messages from a contained object to a container object, and to receive that `Delegate` as a constructor parameter of the contained object. This is exactly the case described above, and it why the constructors of the contained objects must not use those `Delegate` instances from their constructors.

----

In `MapViewState`, there is a `Signal` for when a map change happens. The map change `Signal` is called in `MapViewState::load`. In one of `MapViewState` constructors, we take a saved game to load, which calls `MapViewState::load`, which in turn calls the map change `Signal`, and all it's associated `Delegate` instances. Note that `Signal` will have nothing to iterate over and call if no `connect` calls have been made. That means there's no crash with the current code. However, if we were to convert `MapViewState` like many other classes, and add `Delegate` constructor parameters, which set a `Delegate` for the map change `Signal`, then we very much have the problem described above.

A naive conversion attempt using the pattern established for other classes resulted in the follow crash, with a stack trace generated by `valgrind`:
> ==87072==    at 0x16FE3E: ResearchReport::resetCategorySelection() (ResearchReport.cpp:324)
> ==87072==    by 0x17244A: ResearchReport::fillLists() (ResearchReport.cpp:115)
> ==87072==    by 0x1D7232: MainReportsUiState::clearLists() (MainReportsUiState.cpp:369)
> ==87072==    by 0x1C8C5D: GameState::onMapChange() (GameState.cpp:158)
> ==87072==    by 0x16DE2B: NAS2D::DelegateX<void>::operator()() const (Delegate.h:411)
> ==87072==    by 0x1AF6B0: MapViewState::load(NAS2D::Xml::XmlDocument*) (MapViewStateIO.cpp:325)
> ==87072==    by 0x1BA832: MapViewState::initialize() (MapViewState.cpp:282)
> ==87072==    by 0x1CB036: GameState::initializeGameState() (GameState.cpp:74)
> ==87072==    by 0x1CB116: GameState::initialize() (GameState.cpp:94)
> ==87072==    by 0x200D18: NAS2D::StateManager::setState(NAS2D::State*) (StateManager.cpp:59)
> ==87072==    by 0x13FC16: main (main.cpp:169)

In this case, the `MapViewState` constructor called `load`, which called the `Delegate`, which called `GameState::onMapChange` in the container object, which wasn't yet fully constructed. Curiously, it would have already constructed the contained `MainReportsUiState` field, which is used next, however that class uses multi-stage initialization, and hasn't yet been fully initialized. In particular `MainReportsUiState` is used before the `GameState::initializeGameState()` function is called, which calls `MainReportsUiState::initialize()`, which initializes the `static` `Panels` data, which creates all the `ReportInterface` instances, such as the `ResearchReport`. Hence the crash when using the uninitialized `ResearchReport` object.

**Comment by DanRStevens** [comment]
That completes the conversion of OPHD code from using `Signal` to using `Delegate`. The only remaining uses of `Signal` are indirect uses through the include of `EventHandler` and `MixerSDL`.

The only include of `MixerSDL` is in `main.cpp`. There are many includes for `EventHandler`, such as from `main.cpp`.

There are perhaps a few includes of `EventHandler.h` from header files, which can still be cleaned up.

Related:
- Issue #1573
